### ISSUE TITLE
Accessibility questions review

### ISSUE THREAD

**Issue by dwhly** [hypothesis/client#5788]
We have a customer -- Ohio State University -- who is asking about our progress / performance against a number of accessiblity questions.  

"I hope you’re doing well! I wanted to check in with you regarding the accessibility status of the following items which were slated to be completed during the 2022 year.
- Keyboard shortcuts for highlighting and annotating are not conveyed to the user via instruction.
- When users are using the screen reader to read through a document, they are told an annotation exists when passing over one. However, they cannot jump to the annotation and read it. It is worth noting here that using the annotation pane to read annotations enables users to jump to the point in the document that was annotated.
- When reading annotations in the annotation frame, the user has to go backwards through the DOM to get to the search function.
- On screen readers other than VoiceOver, the A and H keys for annotating and highlighting respectively do not work without entering passthrough mode. This is not an ideal user experience and not obvious to the screen reader users.
- On heavily annotated documents, the start and end annotation messages can make the reading experience very strange and unusable. There is a way to turn off annotations that solves this problem...however, that function should utilize a keyboard command as well so users can enable it on the fly."

These were requests made a while ago that we were supposed to report on.  Some of them look like they might have been addressed.  Others are worded a bit strange and may not be cleanly deliverable.  

If we could do a quick scan of these to see which have been addressed, and which potentially *could* be addressed that would be really helpful.

**Comment by acelaya** [comment]
I have checked all this points. These are my notes:

> Keyboard shortcuts for highlighting and annotating are not conveyed to the user via instruction.

I suppose this means that screen readers do not announce the shortcuts for those actions. If so, this one is not implemented yet.

We do include the shortcut for those actions as part of the button titles, but there are a few things to improve to consider this fully done:

* The button title is only announced when the button is focused. The adder popover is not currently focused when opened, which means its buttons are not the next elements in the tab sequence, and therefore those actions are not announced unless you "click" on some of them.
  I would suggest we focus the adder tooltip after it is opened.
* The shortcut is only added in brakets as part of the button title (for example `Annotate (a)`). This makes screen readers announce it, but it's not super ovbious what the `a` means.
  I would suggest using a title like `Annotate (press a to create a new annotation)`.

> When users are using the screen reader to read through a document, they are told an annotation exists when passing over one. However, they cannot jump to the annotation and read it. It is worth noting here that using the annotation pane to read annotations enables users to jump to the point in the document that was annotated.

This is not implemented yet. There's probably some old issues related with this, but they never became the highest priority.

However, I'm not completely sure what they mean by `passing over one`. Those elements are currently not focusable (that's the main issue to solve), but if they mean hovering over with the mouse, then they would be able to click on them.

> When reading annotations in the annotation frame, the user has to go backwards through the DOM to get to the search function.

I guess they mean they have to `Shift+Tab` (traverse the focus sequence backwards) until the search input is focused.

They can actually open the search from anywhere by pressing `/` or `Cmd+k`/`Ctrl+k` (this was added by Lyza around one year ago).The only limitation to this is that it only works if the sidebar is focused. This is a limitation to our multi-frame architecture, and something we have discussed a couple of times.

> On screen readers other than VoiceOver, the A and H keys for annotating and highlighting respectively do not work without entering passthrough mode. This is not an ideal user experience and not obvious to the screen reader users.

Yes, this is true. Potential solutions:

* Try to change the shortcuts to non-conflicting ones.
* Try to use a combination of keys (maye `Cmd+a` and `Cmd+h` instead of just `a` and `h` could solve this, but something to be tested).
* Check if there's a standard way to let screen readers know about shortcuts so that they announce the need for passthrough mode.

> On heavily annotated documents, the start and end annotation messages can make the reading experience very strange and unusable. There is a way to turn off annotations that solves this problem...however, that function should utilize a keyboard command as well so users can enable it on the fly."

We could add a shortcut to this button. It would be an easy win.

It has the same multi-frame limitation I mentioned above, but in this case, this button is part of the host frame containing the document, so if you are reading over it, it is likely the focused one.

**Comment by robertknight** [comment]
> I suppose this means that screen readers do not announce the shortcuts for those actions. If so, this one is not implemented yet.

Are you sure? I see this code in the client:

https://github.com/hypothesis/client/blob/be0d5b0cb8febe72735b0d6ea518a21e2ecf88eb/src/annotator/components/AdderToolbar.tsx#L117

With the description:

```js
// Render non-visible content for screen readers to announce adder keyboard shortcuts and
// count of annotations associated with the current selection.
```

> When users are using the screen reader to read through a document, they are told an annotation exists when passing over one. However, they cannot jump to the annotation and read it.

If you can select text (eg. via screen reader commands, or any other method), then you can select the highlighted text and press "S" to jump to the annotation in the sidebar and read it. This works in Google Chrome with VoiceOver for example. I believe it works with Microsoft Narrator on Windows but I haven't tested the latest Windows 11 version.

The biggest problem last I worked in this area is that NVDA and JAWS on Windows didn't provide a way to interact with the browser's native selection.

Using a mouse, highlights are clickable. So the most direct way of mapping that to a screen reader would be if the highlight would indicate its clickability.

> On screen readers other than VoiceOver, the A and H keys for annotating and highlighting respectively do not work without entering passthrough mode. This is not an ideal user experience and not obvious to the screen reader users.

I'm not sure how reasonable it is to expect screen reader users to be aware of passthrough mode. There are a lot of major websites that have single-letter shortcuts (YouTube, Gmail, Google and more). The various Google properties show available shortcuts when you press Shift+?. I note that eg. Gmail and YouTube have two sets of shortcuts though, one that use single letter keys and others that use modifiers.

The main issue with using a modifier in Hypothesis is the potential to conflict with either the browser or website's own shortcuts. We might be able to disregard the potential for conflicts for shortcuts which only apply when there is a text selection.

> On heavily annotated documents, the start and end annotation messages can make the reading experience very strange and unusable. 

It would probably make for a nicer experience if we could limit these notices to once per paragraph or something. Some annotation tools similarly don't display visual highlights but instead show a bubble alongside a block of text that contains highlights.

> There is a way to turn off annotations that solves this problem...however, that function should utilize a keyboard command as well so users can enable it on the fly."

This seems like a good recommendation and one that should be easy to implement.

**Comment by acelaya** [comment]
> Are you sure? I see this code in the client

Interesting! Well, I'm not sure anymore 😅

I tested it locally and `orca` did not announce those hints. I may need to debug it a bit further.

**Comment by acelaya** [comment]
On a related note, I think this projects contains the issues originally "reported" https://github.com/orgs/hypothesis/projects/79/views/1

**Comment by robertknight** [comment]
I'm going to close this issue on the basis that we've provided answers to the questions here. Please re-open if you have any follow-up questions.
